### ISSUE TITLE
ValueError in RESTORE

### ISSUE THREAD

**Issue by oesteban** [nipy/dipy#439]
I'm having this error:

```
  File "/usr/local/lib/python2.7/dist-packages/dipy/reconst/dti.py", line 1544, in restore_fit_tensor
    dti_params[vox, :] = start_params
ValueError: could not broadcast input array from shape (7) into shape (12)
```

I'm using 0.8.0dev. I guess that the assignment should be something like: ```dti_params[vox, :7] = start_params```, but I'm not sure.

**Comment by arokem** [comment]
Probably closed through https://github.com/nipy/dipy/pull/498, and related PRs

**Comment by oesteban** [comment]
@arokem, I'll get you some feedback when I test it, but I'm not sure when this is going to happen because I switched my project. Anyways, thanks a lot.

**Comment by joshuacwnewton** [comment]
Sorry to necrobump this issue, but I believe we've run into this error once more (8+ years later! :open_mouth:)

The code has changed quite a bit since 2014, but here is the error we're receiving:

```
  File "/Users/valosek/code/sct_latest/spinalcordtoolbox/scripts/sct_dmri_compute_dti.py", line 150, in compute_dti
    tenfit = dti_restore.fit(data)
  File "/Users/valosek/code/sct_latest/python/envs/venv_sct/lib/python3.8/site-packages/dipy/reconst/dti.py", line 813, in fit
    params_in_mask = self.fit_method(
  File "/Users/valosek/code/sct_latest/python/envs/venv_sct/lib/python3.8/site-packages/dipy/reconst/dti.py", line 1859, in restore_fit_tensor
    D = ols_fit_tensor(design_matrix, flat_data, return_lower_triangular=True)
  File "/Users/valosek/code/sct_latest/python/envs/venv_sct/lib/python3.8/site-packages/dipy/reconst/dti.py", line 1327, in wrapped_fit_tensor
    dtiparams[i:i + step] = fit_tensor(design_matrix,
ValueError: could not broadcast input array from shape (10000,7) into shape (10000,12)
```

The calling code is as follows:

```python
    from dipy.io import read_bvals_bvecs
    bvals, bvecs = read_bvals_bvecs(fname_bvals, fname_bvecs)

    from dipy.core.gradients import gradient_table
    gtab = gradient_table(bvals, bvecs)

    import dipy.denoise.noise_estimate as ne
    sigma = ne.estimate_sigma(data)

    import dipy.reconst.dti as dti
    dti_restore = dti.TensorModel(gtab, fit_method='RESTORE', sigma=sigma)
    tenfit = dti_restore.fit(data)
```

> It would be helpful to me if you could share the data from the voxel on which this is failing. Would that be possible?

`data` in this case is a nifti image read into a numpy array using `nibabel`. Our data is as follows: [data.zip](https://github.com/dipy/dipy/files/12480814/data.zip)

**Comment by arokem** [comment]
Hello! Before I start trying to address the issue, what version of DIPY are you using? 

You can confirm that by running:
```
import dipy
print(dipy.__version__)
```

**Comment by joshuacwnewton** [comment]
Thank you for the response! I have `dipy==1.7.0` installed right now. :)

**Comment by joshuacwnewton** [comment]
I tried looking into the source code to get to the bottom of this issue, and it looks like this issue has actually already been fixed between `dipy==1.7.0` and the current source code on master.

The relevant fix comes from https://github.com/dipy/dipy/pull/2730 (specifically https://github.com/dipy/dipy/commit/50b3d23dc8874f6d138f860d06d35eced00a2d38):

https://github.com/dipy/dipy/blob/796d36620d22eb968b0ef8c5ab03cd8f6116aec2/dipy/reconst/dti.py#L1334

When `return_lower_triangular=True`, then `fit_result` has size (x, 7), which is why the conditional is needed.

By comparison, my local `dipy==1.7.0` source code does not have this line, and instead hardcodes `12` when creating the `dtiparams` matrix:

```python
            dtiparams = np.empty((size, 12), dtype=np.float64)
```

Hence the mismatch error.

So, I guess we'll just have to wait for the next patch release of `dipy`. :)

**Comment by joshuacwnewton** [comment]
Just to comment on one other thing:

> It fails for some of our testing data, but not all of it.

The reason why some of our testing data passed is because those images were smaller than the step size, resulting in an early return with no matrix assignment necessary.
